x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java

```diff
@@ -3,14 +3,847 @@
+            this.encryptionMetadataBlobContainer =
+                    delegatedBlobStore.blobContainer(delegatedBasePath.add(ENCRYPTION_METADATA_ROOT).append(path));
+        }
+
+        /**
+         * Returns the {@link BlobPath} to where the <b>encrypted</b> blobs are stored. Note that the encryption metadata is contained
+         * in separate blobs which are stored under a different blob path (see
+         * {@link #encryptionMetadataBlobContainer}). This blob path resembles the path of the <b>encrypted</b>
+         * blobs but is rooted under a specific path component (see {@link #ENCRYPTION_METADATA_ROOT}). The encryption is transparent
+         * in the sense that the metadata is not exposed by the {@link EncryptedBlobContainer}.
+         *
+         * @return  the BlobPath to where the encrypted blobs are contained
+         */
+        @Override
+        public BlobPath path() {
+            return path;
+        }
+
+        /**
+         * Returns a new {@link InputStream} for the given {@code blobName} that can be used to read the contents of the blob.
+         * The returned {@code InputStream} transparently handles the decryption of the blob contents, by first working out
+         * the blob name of the associated metadata, reading and decrypting the metadata (given the repository password and utilizing
+         * the {@link PasswordBasedEncryption}) and lastly reading and decrypting the data blob, in a streaming fashion by employing the
+         * {@link DecryptionPacketsInputStream}. The {@code DecryptionPacketsInputStream} does not return un-authenticated data.
+         *
+         * @param   blobName
+         *          The name of the blob to get an {@link InputStream} for.
+         */
+        @Override
+        public InputStream readBlob(String blobName) throws IOException {
+            // this requires two concurrent readBlob connections so it's possible that, under lab conditions, the storage service
+            // is saturated only by the first read connection of the pair, so that the second read connection (for the metadata) can not be
+            // fulfilled. In this case the second connection will time-out which will trigger the closing of the first one, therefore
+            // allowing other pair connections to complete. In this situation the restore process should slowly make headway, albeit under
+            // read-timeout exceptions
+            final InputStream encryptedDataInputStream = delegatedBlobContainer.readBlob(blobName);
+            try {
+                // read the metadata identifier (fixed length) which is prepended to the encrypted blob
+                final byte[] metaId = encryptedDataInputStream.readNBytes(MetadataIdentifier.byteLength());
+                if (metaId.length != MetadataIdentifier.byteLength()) {
+                    throw new IOException("Failure to read encrypted blob metadata identifier");
```

Comment:
Maybe make this `<` and make it clear in the exception message that the blob is corrupted/too-short?
(maybe we could technically get here when mounting an unencrypted repository? Should we mention the possibility of that in the exception message?)